pallets/xcm-adapter/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+		debug::info!("------------------------------------------------");
+		Ok(asset.clone())	
+	}
+}
+
+pub struct IsConcreteWithGeneralKey<CurrencyId, FromRelayChainBalance>(
+	PhantomData<(CurrencyId, FromRelayChainBalance)>,
+);
+impl<CurrencyId, B, FromRelayChainBalance> MatchesFungible<B>
+	for IsConcreteWithGeneralKey<CurrencyId, FromRelayChainBalance>
+where
+	CurrencyId: TryFrom<Vec<u8>>,
+	B: TryFrom<u128>,
+	FromRelayChainBalance: Convert<u128, u128>,
+{
+	fn matches_fungible(a: &MultiAsset) -> Option<B> {
+		if let MultiAsset::ConcreteFungible { id, amount } = a {
+			if id == &MultiLocation::X1(Junction::Parent) {
+				// Convert relay chain decimals to local chain
+				let local_amount = FromRelayChainBalance::convert(*amount);
+				return CheckedConversion::checked_from(local_amount);
+			}
+			if let Some(Junction::GeneralKey(key)) = id.last() {
+				if TryInto::<CurrencyId>::try_into(key.clone()).is_ok() {
+					return CheckedConversion::checked_from(*amount);
+				}
+			}
+		}
+		None
+	}
+}
+
+pub trait XCurrencyIdConversion {
+	fn from_asset_and_location(asset: &MultiAsset, location: &MultiLocation) -> Option<PHAXCurrencyId>;
+}
+
+pub struct XCurrencyIdConverter<NativeTokens>(
+	PhantomData<NativeTokens>,
+);
+impl <NativeTokens: Get<BTreeMap<Vec<u8>, MultiLocation>>>  XCurrencyIdConversion for XCurrencyIdConverter<NativeTokens>
+{
+	fn from_asset_and_location(multi_asset: &MultiAsset, multi_location: &MultiLocation) -> Option<PHAXCurrencyId> {
+		if let MultiAsset::ConcreteFungible { ref id, .. } = multi_asset {
+			if id == &MultiLocation::X1(Junction::Parent) {
+				let relaychain_currency : PHAXCurrencyId = PHAXCurrencyId {
+					chain_id: ChainId::RelayChain,
+					currency_id: b"DOT".to_vec(),
+				};
+				return Some(relaychain_currency);
+			}
+
+			if let Some(Junction::GeneralKey(key)) = id.last() {
+				if NativeTokens::get().contains_key(&key.clone()) {
+					// here we can trust the currency matchs the parachain, case NativePalletAssetOr already check this
+					if let MultiLocation::X2(Junction::Parent, Junction::Parachain {id: paraid}) = NativeTokens::get().get(&key.clone()).unwrap() {
+						let parachain_currency: PHAXCurrencyId = PHAXCurrencyId {
+							chain_id: ChainId::ParaChain((*paraid).into()),
+							currency_id: key.clone(),
+						};
+						return Some(parachain_currency);
+					}
+				}
+			}
+		}
+		None
+	}
+}
+
+pub struct NativePalletAssetOr<NativeTokens>(PhantomData<NativeTokens>);
+impl<NativeTokens: Get<BTreeMap<Vec<u8>, MultiLocation>>> FilterAssetLocation for NativePalletAssetOr<NativeTokens> {
+	fn filter_asset_location(asset: &MultiAsset, origin: &MultiLocation) -> bool {
+		if NativeAsset::filter_asset_location(asset, origin) {
+			return true;
+		}
+
+		// native asset identified by a general key
+		if let MultiAsset::ConcreteFungible { ref id, .. } = asset {
+			if let Some(Junction::GeneralKey(key)) = id.last() {
+				if NativeTokens::get().contains_key(&key.clone()) {
```

Comment:
Yeah copy is no necessary here, cctually key is a referenc already, so it can be passed to **contains_key** directly. Same issues would also be fixed.